src/ss.rs

```diff
@@ -0,0 +1,330 @@
+use rand::Rng;
+use sha3::{Digest, Sha3_256};
+
+pub struct BitXORShare {}
+
+impl BitXORShare {
+    #[must_use]
+    pub fn create(value: u32, share_amount: u8) -> Vec<u32> {
+        if share_amount < 2 {
+            panic!("Amount of shares must be at least 2.");
+        }
+        let mut rng = rand::thread_rng();
+        let mut last_sh = value;
+        let mut sh_vec = vec![];
+        for _i in 0..share_amount - 1 {
+            let sh = rng.gen::<u32>();
+            sh_vec.push(sh);
+            last_sh ^= sh;
+        }
+        sh_vec.push(last_sh);
+        sh_vec
+    }
+
+    pub fn combine(shares: Vec<u32>) -> u32 {
+        let mut value = 0;
+        for sh in shares {
+            if value == 0 {
+                value = sh;
+            } else {
+                value ^= sh;
+            }
+        }
+        value
+    }
+}
+
+pub struct SecureCoinTossing {}
+
+impl SecureCoinTossing {
+    pub fn get_random_and_hash() -> (u128, String) {
+        // generate a u128 random integer.
+        let mut rng = rand::thread_rng();
+        let r = rng.gen::<u128>();
+
+        // generate hash for the random integer's bit string.
+        let mut hasher = Sha3_256::new();
+        hasher.update(format!("{:b}", r));
+        let h: String = format!("{:X}", hasher.finalize());
+        (r, h)
+    }
+
+    pub fn get_random() -> u128 {
+        // generate a u128 random integer.
+        let mut rng = rand::thread_rng();
+        rng.gen::<u128>()
+    }
+
+    pub fn verify_random_and_hash(r: u128, h: String) -> bool {
+        // generate hash for the random integer's bit string.
+        let mut hasher = Sha3_256::new();
+        hasher.update(format!("{:b}", r));
+        let hv: String = format!("{:X}", hasher.finalize());
+        h == hv
+    }
+
+    pub fn calculate_shared_random(r1: u128, r2: u128) -> u128 {
+        r1 ^ r2
+    }
+}
+
+pub struct RejectionSampling {}
+
+impl RejectionSampling {
+    pub fn generate_permutation(bit_string_as_int: u128, dimension: usize) -> Vec<usize> {
```

Comment:
Your source of entropy is a little scant here.  You probably want `&[u8]` rather than `u128` for that.  Or `rand::Rng`.  Note that your use of rejection sampling will ensure that you will run out of bits if you provide a fixed-length input of any size; you really want an RNG (or really a PRF with a counter that can be extended arbitrarily).